### PR TITLE
openssl v3.0.0

### PR DESCRIPTION
**Pull request by h-vetinari**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

-----

Some big changes with 3.0.0, see [announcement](https://www.openssl.org/blog/blog/2021/09/07/OpenSSL3.Final/) and [migration guide](https://www.openssl.org/docs/man3.0/man7/migration_guide.html).

Some TODOs (no claim about exhaustiveness):
* [x] add `enable-fips` unconditionally (or use a different output)?
* [x] add `enable-ktls` on linux?
* [x] set `OPENSSL_VERSION_BUILD_METADATA=conda_forge` (see [here](https://www.openssl.org/docs/man3.0/man3/OpenSSL_version.html))?
* [x] drop patch for `USE_BCRYPTGENRANDOM` plus obsolete VS-versions?
* [x] ~other build changes necessary for providers~ (none 🥳)
* [x] clean up skips & test suite validation
* [x] make perl build-only dep again

### CODE REVIEW

**Comment by h-vetinari** [comment]
Lots of errors for the new CMP implementation under unix:
```
Test Summary Report
-------------------
80-test_cmp_http.t               (Wstat: 1280 Tests: 5 Failed: 5)
  Failed tests:  1-5
  Non-zero exit status: 5
Files=241, Tests=3274, 418 wallclock secs ( 7.09 usr  0.69 sys + 358.41 cusr 59.86 csys = 426.05 CPU)
```

**Comment by h-vetinari** [comment]
@conda-forge/core, this is failing in some new parts of OpenSSL, where the tests spin up a HTTP server, and somehow aren't able to reach that within CI.

From the upstream issue:
> The test server startup succeeds and the client tries connecting to the right port, but very likely it cannot reach the server via the local IPv4 address 127.0.0.1

Are there any ways that conda (or azure) block off 127.0.0.1? Or should we just skip `80-test_cmp_http.t`? Everything else passes.

File: recipe/build.sh [hunk 1]
```diff
@@ -37,34 +33,18 @@ case "$target_platform" in
     ;;
 esac
 
+# https://github.com/openssl/openssl/issues/16546
+if [ "${}" = "linux-"* ]; then
+  sed -i "s@127.0.0.1@$(hostname -I)@g" test/recipes/80-test_cmp_http.t test/recipes/80-test_cmp_http_data/Mock/test.cnf
```
Comment (h-vetinari) [inline recipe/build.sh:38]: Thanks for helping out here @chrisburr!

I think this is going to need to apply to both linux & osx.

Comment (chrisburr) [reply recipe/build.sh]: `hostname -I` doesn't work on macOS, I was hoping to grep it from `ifconfig`

**Comment by h-vetinari** [comment]
@chrisburr 
I took the liberty of taking your idea and running with it; clearly, the hostname on azure is something different, but also the sed-replacement seems to  be inserting a spurious space before the port:
```
# cmp_main:apps/cmp.c:2668:CMP info: using section(s) 'Mock connection' of OpenSSL configuration file '../Mock/test.cnf'
../../../../util/wrap.pl ../../../../apps/openssl cmp -server '172.17.0.2 :39227' -config ../Mock/test.cnf -section 'Mock connection' -certout ../../../../test-runs/test_cmp_http/test.cert.pem -proxy => 1
    not ok 1 - default config
```

No idea yet why...

File: recipe/build.sh (hunk 1 above)
Comment (h-vetinari) [reply recipe/build.sh]: Ah, apologies, will reinstate with the next push.

**Comment by h-vetinari** [comment]
By setting `HOST_IP` separately, the space is gone, but the tests still fail:
```
../../../../util/wrap.pl ../../../../apps/openssl cmp -server '172.17.0.2:40747' -config ../Mock/test.cnf -section 'Mock connection' -certout ../../../../test-runs/test_cmp_http/test.cert.pem -proxy -no_proxy 172.17.0.2 => 1
    not ok 1 - default config
```

**Comment by h-vetinari** [comment]
This was the output of `ifconfig` on osx:
```
lo0: flags=8049<UP,LOOPBACK,RUNNING,MULTICAST> mtu 16384
	options=1203<RXCSUM,TXCSUM,TXSTATUS,SW_TIMESTAMP>
	inet 127.0.0.1 netmask 0xff000000 
	inet6 ::1 prefixlen 128 
	inet6 fe80::1%lo0 prefixlen 64 scopeid 0x1 
	nd6 options=201<PERFORMNUD,DAD>
gif0: flags=8010<POINTOPOINT,MULTICAST> mtu 1280
stf0: flags=0<> mtu 1280
en0: flags=8863<UP,BROADCAST,SMART,RUNNING,SIMPLEX,MULTICAST> mtu 1500
	options=40b<RXCSUM,TXCSUM,VLAN_HWTAGGING,CHANNEL_IO>
	ether 00:50:56:ac:c9:36 
	inet 10.212.8.92 netmask 0xffffff80 broadcast 10.212.8.127
	media: autoselect (1000baseT <full-duplex>)
	status: active
utun0: flags=8051<UP,POINTOPOINT,RUNNING,MULTICAST> mtu 1380
	inet6 fe80::4bda:9670:b671:f98e%utun0 prefixlen 64 scopeid 0x5 
	nd6 options=201<PERFORMNUD,DAD>
utun1: flags=8051<UP,POINTOPOINT,RUNNING,MULTICAST> mtu 2000
	inet6 fe80::99dc:2ded:b7d7:2365%utun1 prefixlen 64 scopeid 0x6 
	nd6 options=201<PERFORMNUD,DAD>
```

**Comment by chrisburr** [comment]
There is something odd very here, I was able to reproduce the failure locally and with the fix it works for me with `conda debug` but not `conda build`...

**Comment by h-vetinari** [comment]
Upstream not(ic)ed, that this was due to having `HTTP_PROXY` set to an empty string, apparently. Unsetting the variable fixed the unix builds.

This now builds cleanly, but IMO the following points should still be decided upon:
* [ ] add `enable-fips` unconditionally (or use a different output)?
* [ ] add `enable-ktls` on linux?
* [ ] set `OPENSSL_VERSION_BUILD_METADATA=conda_forge` (see [here](https://www.openssl.org/docs/man3.0/man3/OpenSSL_version.html))?
* [ ] drop patch for `USE_BCRYPTGENRANDOM` plus obsolete VS-versions?

**Comment by h-vetinari** [comment]
> This now builds cleanly, but IMO the following points should still be decided upon:

Since there was no response (other than hearts), I went ahead and did the 4 things that I had noted as to-be-decided. 🙃 

Please review @conda-forge/openssl

File: recipe/meta.yaml [hunk 2]
```diff
@@ -1,57 +1,45 @@
...
-    # For "perl" in "requirements/host" on windows, see below.
-    - perl *
 
 requirements:
   build:
     - {{ compiler('c') }}
     - nasm               # [win]
     - make               # [unix]
-    - perl *             # [not win]
+    - perl
```
Comment (isuruf) [inline recipe/meta.yaml:23]:
```suggestion
    - perl *
```

File: recipe/meta.yaml [hunk 3]
```diff
@@ -1,57 +1,45 @@
...
-        - perl *
+        # openssl's versioning used to be X.Y.Z(rev), but is MAJOR.MINOR.PATCH since 3.0.0;
+        # see https://www.openssl.org/policies/releasestrat.html
+        - {{ pin_subpackage('openssl', max_pin='x') }}
     requirements:
       build:
         - {{ compiler('c') }}
         - make
-        - perl *             # [not win]
+        - perl
```
Comment (isuruf) [inline recipe/meta.yaml:40]:
```suggestion
        - perl *
```

**Review by isuruf — APPROVED** [review]

File: recipe/meta.yaml (hunk 2 above)
Comment (h-vetinari) [reply recipe/meta.yaml:23]: Is there a functional difference between the two (or anything worth indicating with the `*`)? If yes, I was unaware & can re-add of course.

Comment (isuruf) [reply recipe/meta.yaml:23]: Yes, if there's no `*`, conda-build will generate multiple jobs if there was multiple versions for `perl` in conda-forge pinning.

Comment (h-vetinari) [reply recipe/meta.yaml:23]: Ah, makes sense, thx.

**Review by chrisburr — APPROVED** [review]

**Pull request merged**